### PR TITLE
Untested: libconsensus: Expose VerifyHeader

### PR DESCRIPTION
**Pull request by jtimon**
VerifyHeader basically calls CheckBlockHeader and ContextualCheckBlockHeader for a given CBlockHeader (and all the required parameters).

Since ContextualCheckBlockHeader previously depends on CBlockIndex, some kind of C-compatible API for it must be exposed if we want a complete libconsensus to remain storage agnostic. 
Alternative implementations of BlockIndexInterface are more than welcomed.
If there's parts of this PR that you think would be acceptable separately, please, say so.
If you come up with alternative preparations for making this PR (or an alternative) smaller, that's welcomed too. And nits, if I haven't said so before, are always welcomed.

 We could also expose CheckBlockHeader independently. An SPV wallet using libconsensus for verifyscript already would bea easily able to also easily check CheckTransaction (which @NicolasDorier is working on) and CheckBlockHeader without any context. If we do so, we could easily move the time checks from ContextualCheckBlockHeader to CheckBlockHeader if we accept the current time of the node as a "non-contextual" parameter for CheckBlockHeader. Obviously time is context, but being pragmatic, these checks are to easy not to expose in CheckBlockHeader and everybody can be expected to have a clock they can trust (the error margins in your clock seem quite safe to me and, maybe more importante, VerifyHeader is going to use the caller's clock anyway).

Please feel free to ask for squashes as well.

Replaces #5995

Depdendencies:
- [ ] Consensus: MOVEONLY: Move functions for header verification #8337
- [x] finish implementing bitcoinconsensus_create_consensus_parameters (and remove whatever is not needed anymore from the base_blob<> preparations).

Death TODOs:

~~\- Hide uint256 and DifficultyAdjustmentInterval() in Consensus::Params~~: the create interface should be enough.
~~\- [ ] The interface for CBlockIndex (`const BlockIndexInterface* iBlockIndex`) could be hidden in libconsensus using the same create/destroy blockindexInterfaceTrick, even if some of the parameters are going to be function pointers. Or we could offer both and see what happens...discussion welcomed here, but allowing BlockIndexInterface\* to be passed as void\* it's in the todo list whether for necessary or just to see how it looks like. If we ever need a DifficultyAdjustmentInterval-like method internally, void\* will be transparent. It's simply the same argument and we should do the same.~~
~~\- [ ] new warnings~~

### CODE REVIEW

**Comment by luke-jr** [comment]
This API doesn't look C-compatible. Am I missing something?

Dislike replacing uint256 with char arrays, but I'm not sure what the best alternative is; is it possible to make uint256 a C-compatible struct?

**Comment by jtimon** [comment]
I guess a struct or a typedef would work.

What do you think is not C compatible? What am I missing?

**Comment by luke-jr** [comment]
consensus/consensus.h isn't currently exposed in libconsensus, but namespaces, the & operator, and classes won't work in C. The function also needs to be wrapped in an extern "C" block to get the right ABI.

**Comment by jtimon** [comment]
consensus/consensus.h is not to be exposed in libconsensus.
I didn't knew & wasn't C, but that can be trivially changed to pointers.
I didn't knew namespaces weren't C either. Consensus::Params was a big
mistake then...talking it out of the namespace is going to be disruptive.

**Comment by sipa** [comment]
Please don't go turn the entire codebase into C just because we want a
C-capable API.

You can create a wrapper object in C that just contains a pointer to a
malloced Consensus::Params. You could have a
libconsensus_params_create(...) function that calls new Consensus::Params()
and sets some fields in it, and return it cast to a void*. The caller would
be required to call libconsensus_params_destroy() afterwards to clean it up.

**Comment by jtimon** [comment]
Yes, or just recreating the struct without the namespace and convert it
inside the exposed function. We can slowly replace Consensus::Params for
ConsensusParams or whatever later.
Still, had I known namespaces weren't C, ConsensusParams could have lacked
a namespace from the beginning and this wrapping wouldn't be necessary.

**Comment by sipa** [comment]
Yes it would be.

Consensus::Params is a neatly encapsulated C++ class, as it should be. It
has private fields and getters, and we're free to change its internal
representation from time to time (and we have done so many times).

Converting it to a C struct with a stable representation would be an
incredible burden if we'd want to start adding extra fields. It being in a
namespace doesn't matter for this at all.

Please... create a wrapper that constructs C++ objects for the C interface.
Not doing so would very badly limit the possible changes we can make later.

**Comment by jtimon** [comment]
Yes, there's no other clean way now.
I still lament not having even followed my own plan of keeping Consensus::Params as C-compatible as possible...by forgetting namespaces (really, thanks @luke-jr for pointing that out).

But yeah, I've previously discussed with @NicolasDorier that we could wrap all fields in something like struct BlockIndexInterface https://github.com/bitcoin/bitcoin/pull/8493/commits/fb9692bc74a38233c082d90566051a5ea7eb6537
around set_pointer_function_for_pseudo_method_x() style a interface and that would work similarly to not wrapping it with some small advantages even if we wanted to have the inside written in C, which I don't realistically think about anymore in any case.

In this case, as you say, it is as easy as two functions: create_params() and destroy_params().

Thanks @luke-jr @sipa @NicolasDorier for this tremendously useful feedback.

File: src/script/bitcoinconsensus.h [hunk 1]
```diff
@@ -6,6 +6,9 @@
 #ifndef BITCOIN_BITCOINCONSENSUS_H
 #define BITCOIN_BITCOINCONSENSUS_H
 
+#include "consensus/interfaces.h"
+#include "consensus/params.h"
+
```
Comment (jtimon) [inline src/script/bitcoinconsensus.h]: NACK these two includes.

**Comment by jtimon** [comment]
Rebased on top of #8337. For convenience, one can see the changes without the moveonly in: https://github.com/jtimon/bitcoin/compare/0.12.99-consensus-moveonly-header...jtimon:0.13-consensus-header

As discussed:

- consensus/params.h is no longer included by script/bitcoinconsensus.h: bitcoinconsensus_create_consensus_parameters and bitcoinconsensus_destroy_consensus_parameters are created instead.

- This also saves all the preparations for using a 32 byte array instead of a uint256 in Consensus::Params since its only going to be exposed in the create function: no need to disrupt any of the 3 hashes in Consensus::Params nor the convenience method. Quite a few diffs have been saved here, thanks again.

The TODOs will be updated at the end of the OP, but here's the current summary:

- finish implementing bitcoinconsensus_create_consensus_parameters (and remove whatever is not needed anymore from the base_blob<> preparations).

- The interface for CBlockIndex (```const BlockIndexInterface* iBlockIndex```) could be hidden in libconsensus using the same create/destroy blockindexInterfaceTrick, even if some of the parameters are going to be function pointers. Or we could offer both and see what happens...discussion welcomed here, but allowing BlockIndexInterface* to be passed as void* it's in the todo list whether for necessary or just to see how it looks like.

- new warnings

**Comment by jtimon** [comment]
Added some commits fixing some nits I received on IRC. Didn't squashed in case someone doesn't like the changes.

**Comment by jtimon** [comment]
Squashed, should be almost-identical to previous de7a2bb but with a cleaner history.

Short term next steps:

- Unify consensus/functionptr and consensus/interfaces
- Create header_verify.h instead of declaring the functions in consensus/consensus.h 
- Pay attention to further feedback

**Comment by jtimon** [comment]
Unify consensus/functionptr and consensus/interfaces breaks previous da18576

**Comment by jtimon** [comment]
More changes on top of 9628c76 : 
- [X] Unify consensus/functionptr and consensus/interfaces
- [X] Create header_verify.h instead of declaring the functions in consensus/consensus.h
Also, rebased and some other minors fixes.

**Comment by jtimon** [comment]
Needed rebase after renaming main.o, see #9260
The needed rebase was clean modulo fixed nits in #8337 .
Reminder: nits and reasoned or questioning NACKs welcomed.

**Comment by jtimon** [comment]
This has been needing rebase for a while. Closing for lack of interest.

**Pull request closed without merging**